pkg/ruler/base/ruler.go

```diff
@@ -422,6 +454,17 @@ func instanceOwnsRuleGroup(r ring.ReadRing, g *rulespb.RuleGroupDesc, instanceAd
 	return rlrs.Instances[0].Addr == instanceAddr, nil
 }
 
+func instanceOwnsRule(r ring.ReadRing, rg *rulespb.RuleGroupDesc, rd *rulespb.RuleDesc, instanceAddr string) (bool, error) {
+	hash := tokenForRule(rg, rd)
+
+	rlrs, err := r.Get(hash, RingOp, nil, nil, nil)
+	if err != nil {
```

Comment:
From here onwards within this function, it's the same as `instanceOwnsRuleGroup`. I'd extract this logic to a new function and call it from both `instanceOwnsRuleGroup` and `instanceOwnsRuleGroup`.

Reply:
I want to keep them separate because they might diverge in future; we don't save much by abstracting this IMHO